ex_cubic_ods_ingestion/lib/ex_cubic_ods_ingestion/start_ingestion.ex

```diff
@@ -0,0 +1,124 @@
+defmodule ExCubicOdsIngestion.StartIngestion do
+  @moduledoc """
+  StartIngestion module.
+  """
+
+  use GenServer
+
+  alias ExCubicOdsIngestion.ProcessIngestion
+  alias ExCubicOdsIngestion.Schema.CubicOdsLoad
+  alias ExCubicOdsIngestion.Schema.CubicOdsTable
+  alias ExCubicOdsIngestion.Workers.Ingest
+
+  require Logger
+  require Oban
+
+  @wait_interval_ms 5_000
+
+  defstruct [:lib_ex_aws, status: :not_started, continuation_token: "", max_keys: 1_000]
+
+  # client methods
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(opts) do
+    # define lib_ex_aws, unless it's already defined
+    opts = Keyword.put_new(opts, :lib_ex_aws, ExAws)
+
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @spec status(GenServer.server()) :: :running
+  def status(server) do
+    GenServer.call(server, :status)
+  end
+
+  # callbacks
+  @impl GenServer
+  def init(opts) do
+    # construct state
+    state = struct!(__MODULE__, opts)
+
+    {:ok, %{state | status: :running}, 0}
+  end
+
+  @impl GenServer
+  def handle_info(:timeout, %{} = state) do
+    new_state = run(state)
+
+    # set timeout according to need for continuing
+    timeout = @wait_interval_ms
+
+    {:noreply, new_state, timeout}
+  end
+
+  @impl GenServer
+  def handle_call(:status, _from, state) do
+    {:reply, state.status, state}
+  end
+
+  # server helper functions
+  @spec run(map()) :: map()
+  defp run(state) do
+    # get list of load records that are in 'ready' state, ordered by s3_modified, s3_key
+    ready_load_recs = CubicOdsLoad.get_status_ready()
+
+    # attach table to load records, and update snapshots
+    ready_loads = Enum.map(ready_load_recs, &attach_table(&1))
+
+    # iterate through the load records list in order to start ingesting job,
+    # or error
+    Enum.each(ready_loads, &start_ingestion(&1))
+
+    # return
+    state
+  end
+
+  @spec attach_table(CubicOdsLoad.t()) :: tuple()
+  def attach_table(load_rec) do
+    # find the table rec that the load is for
+    table_rec =
+      if load_rec.table_id do
+        CubicOdsTable.get(load_rec.table_id)
+      else
+        CubicOdsTable.get_from_load_s3_key(load_rec.s3_key)
+      end
+
+    if table_rec do
+      # for table, update snapshot if we have a new snapshot, as in
+      # the load key matches the snapshot key and the load modified date is newer than the snapshot
+      table_rec =
+        if table_rec.snapshot_s3_key == load_rec.s3_key and
+             table_rec.snapshot < load_rec.s3_modified do
```

Comment:
suggestion: `s3_modified` and `snapshot` are both DateTimes, which mean you can't compare them this way:

```elixir
iex> ~D[2021-04-30] > ~D[2021-05-01]
warning: invalid comparison with struct literal ~D[2021-04-30]. Comparison operators (>, <, >=, <=, min, and 
max) perform structural and not semantic comparison. Comparing with a struct literal is unlikely to give a 
meaningful result. Struct modules typically define a compare/2 function that can be used for semantic 
comparison
  iex:1

true
```
```suggestion
             DateTime.compare(table_rec.snapshot, load_rec.s3_modified) == :lt do
```

Reply:
You could also try something like `CubicOdsTable.upsert_snapshot(table_rec, load_rec.s3_modified)` and have the DB handle the check about whether the snapshot needs to be changed.

Reply:
Thanks. Oversight on my part. Have updated it. I didn't get this warning you have here on `iex`. Can you let me know how I make these things show up?

Reply:
I don't think I did anything in particular. If you paste in that comparison into IEx, what do you see?

Reply:
Ah. When I try your example, it does fails with that warning. I wasn't getting a warning for the snippet below, which I didn't think was too far from your example:
```ex
prev = DateTime.utc_now()
Process.sleep(1000)
next = DateTime.utc_now()
prev < next
# ... result
true
``